src/analysis/souffle/taint.dl

```diff
@@ -71,6 +95,37 @@ mayHaveTag(tgt, owner, tag) :- hasTag(tgt, owner, tag).
 mayHaveTag(tgt, owner, tag) :-
     resolvedEdge(owner, src, tgt), mayHaveTag(src, owner, tag), !removeTag(tgt, owner, tag).
 
+// Integrity tag rules.
+
+isIntegTag(integTag) :- directHasIntegTag(_, integTag).
+
+// The inverse integrity tag is the same text as the integrity tag except with
+// a '!' in front of it.
+integTagToInv(integTag, cat("!", integTag)) :- isIntegTag(integTag).
+
+// Introduce an inverse integrity tag at any access point that has no incoming
+// edges and is not directly given the corresponding integrity tag.
+mayHaveInvIntegTag(ap, invIntegTag) :-
+  isAccessPath(ap), integTagToInv(integTag, invIntegTag), !edge(_, ap),
+  !directHasIntegTag(ap, integTag).
+
+// Introduce inverse integrity tag wherever a corresponding integrity tag is
+// removed.
+mayHaveInvIntegTag(ap, invIntegTag) :-
+  isAccessPath(ap), integTagToInv(integTag, invIntegTag),
+  removeIntegTag(ap, integTag).
```

Comment:
I find it a bit more intuitive to handle removal of integrity tags in the rule for `mustHaveIntegTag` directly rather than using the removal to introduce an inverse integ tag (or to populate the lackOfIntegTag relation).

Reply:
I've removed the `InvIntegTag` entirely. However, we have to make `removeIntegrityTag` introduce the lack of an `IntegrityTag` rather than directly influence `mustHaveIntegrityTag` to preserve stratification (as `mustHaveIntegrityTag` uses `!lacksIntegrityTag on the RHS of one of its rules).